rasa/server.py

```diff
@@ -476,6 +479,124 @@ async def hello(request: Request):
         return response.text("Hello from Rasa: " + rasa.__version__)
 
 
+def async_if_callback_url(f: Callable[..., Coroutine]) -> Callable:
+    """Decorator to enable async request handling.
+
+    If the incoming HTTP request specified a `callback_url` query parameter, the request
+    will return immediately with a 204 while the actual request response will
+    be sent to the `callback_url`. If an error happens, the error payload will also
+    be sent to the `callback_url`.
+
+    Args:
+        f: The request handler function which should be decorated.
+
+    Returns:
+        The decorated function.
+    """
+
+    @wraps(f)
+    async def decorated_function(
+        request: Request, *args: Any, **kwargs: Any
+    ) -> HTTPResponse:
+        callback_url = request.args.get("callback_url")
+        # Only process request asynchronously if the user specified a `callback_url`
+        # query parameter.
+        if not callback_url:
+            return await f(request, *args, **kwargs)
+
+        async def wrapped() -> None:
+            try:
+                result: HTTPResponse = await f(request, *args, **kwargs)
+                payload = dict(
+                    data=result.body, headers={"Content-Type": result.content_type}
+                )
+                logger.debug(
+                    "Asynchronous processing of request was successful. "
+                    "Sending result to callback URL."
+                )
+
+            except Exception as e:
+                if not isinstance(e, ErrorResponse):
+                    logger.error(e)
+                    e = ErrorResponse(
+                        HTTPStatus.INTERNAL_SERVER_ERROR,
+                        "UnexpectedError",
+                        f"An unexpected error occurred. Error: {e}",
+                    )
+                # If an error happens, we sent the error payload to the `callback_url`
+                payload = dict(json=e.error_info)
+                logger.debug(
+                    "Error happened when processing request asynchronously. "
+                    "Sending error to callback URL."
+                )
+            async with aiohttp.ClientSession() as session:
+                await session.post(callback_url, raise_for_status=True, **payload)
+
+        # Run the request in the background on the event loop
+        request.app.add_task(wrapped())
+
+        # The incoming request will return immediately with a 204
+        return response.empty()
+
+    return decorated_function
+
+
+def heavy_computation(f: Callable[..., Coroutine]) -> Callable:
```

Comment:
This name feels ambiguous to me, better I think to give it a name which describes what it does rather than what it's used for.